tests: drivers: can: api: add testcase for testing CANXL non RX FIFO

congnguyenhuu opened the pull request:
Add testcase for testing CANXL non RX FIFO on s32z270dc2 board

```
python3 scripts/twister -p s32z270dc2_rtu0_r52  --disable-warnings-as-errors --device-testing --device-serial=/dev/ttyUSB0 -v --west-runner trace32 --west-flash="--config=/opt/tools/trace32/config.t32" -T tests/drivers/can/api/
INFO    - Using Ninja..
INFO    - Zephyr version: 0d9bb76dd32f
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...

Device testing on:

| Platform            | ID   | Serial device   |
|---------------------|------|-----------------|
| s32z270dc2_rtu0_r52 |      | /dev/ttyUSB0    |

INFO    - JOBS: 4
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 2/4 s32z270dc2_rtu0_r52       tests/drivers/can/api/drivers.can.api.rtr          PASSED (device 18.347s)
INFO    - 3/4 s32z270dc2_rtu0_r52       tests/drivers/can/api/drivers.can.api              PASSED (device 17.929s)
INFO    - 4/4 s32z270dc2_rtu0_r52       tests/drivers/can/api/drivers.can.api.nxp_s32_canxl.non_rx_fifo PASSED (device 18.107s)

INFO    - 4 test scenarios (4 test instances) selected, 1 configurations skipped (1 by static filter, 0 at runtime).
INFO    - 3 of 4 test configurations passed (100.00%), 0 failed, 0 errored, 1 skipped with 0 warnings in 249.57 seconds
INFO    - In total 121 test cases were executed, 58 skipped on 1 out of total 673 platforms (0.15%)
INFO    - 3 test configurations executed on platforms, 0 test configurations were only built.

```

congnguyenhuu committed:
tests: drivers: can: api: add testcase for testing CANXL non RX FIFO

Add testcase for testing CANXL non RX FIFO

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
tests: drivers: can: api: wrap the code belong to CAN FD mode

To avoid build warnings when building non CAN FD mode

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu commented:
I added a commit to wrap the code belongs to CAN FD mode to avoid build warnings when `CONFIG_CAN_FD_MODE=n`
The build warning:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/76719459/580e5b7d-f5ae-4d73-aa75-97de5bd3462f)

henrikbrixandersen commented on tests/drivers/can/api/testcase.yaml:
```diff
@@ -18,3 +16,13 @@ tests:
       - xiao_esp32s3
+  drivers.can.api.nxp_s32_canxl_non_rx_fifo:
+    extra_configs:
+      - CONFIG_CAN_NXP_S32_RX_FIFO=n
+      - CONFIG_CAN_FD_MODE=n
```
If `CONFIG_CAN_NXP_S32_RX_FIFO=n` does not pass these tests with `CONFIG_CAN_FD_MODE=y`, then the S32 CANXL driver needs to be fixed. 

From the looks of it, the driver needs to treat `CONFIG_CAN_NXP_S32_RX_FIFO=n` just like it treats `CONFIG_CAN_FD_MODE=n` by excluding all CAN FD functionality if the Rx FIFO is used.

congnguyenhuu replied:
This is limitation of CANXL driver after removing CAN FD filtering.
Actually, it still works with `CONFIG_CAN_NXP_S32_RX_FIFO=n` and `CONFIG_CAN_FD_MODE=y` but it just supports to receive CAN classic frame. So it does not pass these tests and I disable CAN FD to skip these.

henrikbrixandersen replied:
The CAN FD test are skipped if the driver does not have the `CAN_MODE_FD` capability. This capability is reported as supported by the NXP S32 CANXL driver even for `CONFIG_CAN_NXP_S32_RX_FIFO=n`. This does not look right.

congnguyenhuu replied:
yes, that's right. CANXL driver is supporting CAN FD capability even for `CONFIG_CAN_NXP_S32_RX_FIFO=n`. I will open other PR to removed CAN FD support for non-RX_FIFO.

henrikbrixandersen replied:
Just fix it in this PR?

henrikbrixandersen replied:
No, please leave the guards in. They are correct even if this test will not be built without CAN FD support.

henrikbrixandersen commented on tests/drivers/can/api/src/common.c:
```diff
@@ -107,6 +108,7 @@ const struct can_frame test_std_fdf_frame_2 = {
 		    46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60,
 		    61, 62, 63, 64 }
 };
+#endif
```
```suggestion
#endif /* CONFIG_CAN_FD_MODE */
```

henrikbrixandersen commented on tests/drivers/can/api/src/common.h:
```diff
@@ -95,6 +96,7 @@ extern const struct can_frame test_std_fdf_frame_1;
  * @brief Standard (11-bit) CAN ID frame 2 with CAN FD payload.
  */
 extern const struct can_frame test_std_fdf_frame_2;
+#endif
```
```suggestion
#endif /* CONFIG_CAN_FD_MODE */
```

henrikbrixandersen commented on tests/drivers/can/api/src/utilities.c:
```diff
@@ -119,9 +119,11 @@ ZTEST(can_utilities, test_can_frame_matches_filter)
 	/* CAN FD format frames and filters */
 	zassert_true(can_frame_matches_filter(&test_std_fdf_frame_1, &test_std_filter_1));
 	zassert_true(can_frame_matches_filter(&test_std_fdf_frame_2, &test_std_filter_2));
+#endif
```
```suggestion
#endif /* CONFIG_CAN_FD_MODE */
```

henrikbrixandersen commented on tests/drivers/can/api/testcase.yaml:
```diff
@@ -18,3 +16,13 @@ tests:
       - esp32s2_saola
       - esp32s3_devkitm
       - xiao_esp32s3
+  drivers.can.api.nxp_s32_canxl_non_rx_fifo:
```
```suggestion
  drivers.can.api.nxp_s32_canxl.non_rx_fifo:
```

henrikbrixandersen commented on tests/drivers/can/api/testcase.yaml:
```diff
@@ -18,3 +16,13 @@ tests:
+    extra_configs:
+      - CONFIG_CAN_NXP_S32_RX_FIFO=n
+      - CONFIG_CAN_FD_MODE=n
+    filter: dt_compat_enabled("nxp,s32-canxl")
+    platform_allow:
+      - s32z270dc2_rtu0_r52
+      - s32z270dc2_rtu1_r52
+      - s32z270dc2_rtu0_r52@D
+      - s32z270dc2_rtu1_r52@D
```
Why both a `filter` and `platform_allow`?

congnguyenhuu replied:
I think need both them to make sure to have a dts canxl node is enabled

congnguyenhuu replied:
yes, it is not needed to use both. I removed `platform_allow`

congnguyenhuu committed:
drivers: nxp_s32_canxl: remove support CAN FD mode for non-RX_FIFO

This is driver limitation after removing CAN_FILTER_FDF flag #65108.
CANXL driver need to know CAN_FILTER_FDF for configuring Rx filter
so that it receives CAN classic or CAN FD frames when using non RX_FIFO.
So update driver that just supports CAN classic for non RX_FIFO.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -152,7 +152,7 @@ static int can_nxp_s32_get_capabilities(const struct device *dev, can_mode_t *ca
 	*cap = CAN_MODE_NORMAL | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY;
 
-#if CONFIG_CAN_FD_MODE
+#if defined(CONFIG_CAN_FD_MODE) && defined(CONFIG_CAN_NXP_S32_RX_FIFO)
```
This will leave a lot of unused code paths compiled into this driver when either `CONFIG_CAN_FD_MODE` or `CONFIG_CAN_NXP_S32_RX_FIFO` is enabled, but the other one isn't.

Wouldn't it make more sense to determine whether this driver supports CAN FD at all (depends on `CONFIG_CAN_FD_MODE=y` && `CONFIG_CAN_NXP_S32_RX_FIFO=y`) at the very top of the file? Create a new define for this and replace all in-driver use of `#ifdef CONFIG_CAN_FD_MODE` with this?

That way, if either `CONFIG_CAN_FD_MODE=n` or `CONFIG_CAN_NXP_S32_RX_FIFO=n`, the driver will be compiled to only support classic CAN and no unused code paths will be included.

congnguyenhuu replied:
Thanks for nice suggestion. I updated this.

henrikbrixandersen commented on drivers/can/can_nxp_s32_canxl.c:
```diff
@@ -50,6 +50,10 @@
 #endif
 
+#if defined(CONFIG_CAN_FD_MODE) && defined(CONFIG_CAN_NXP_S32_RX_FIFO)
+#define CAN_NXP_S32_FD_MODE TRUE
```
```suggestion
#define CAN_NXP_S32_FD_MODE 1
```

congnguyenhuu replied:
I updated

henrikbrixandersen reviewed (changes requested):
Looks good, thanks. One minor nit:
